sixtyfps_compiler/generator.rs

```diff
@@ -143,51 +143,63 @@ pub fn build_array_helper(component: &Component, mut visit_item: impl FnMut(&Ele
     }
 }
 
+/// A reference to this trait is passed to the [`build_item_tree`] function.
+/// It can be used to build the array for the item tree.
+pub trait ItemTreeBuilder {
+    /// Some state that contains the code on how to access some particular component
+    type SubComponentStack;
+
+    fn push_repeated_item(
+        &mut self,
+        item: &crate::object_tree::ElementRc,
+        repeater_count: u32,
+        parent_index: u32,
+        component_stack: &Self::SubComponentStack,
+    );
+    fn push_native_item(
+        &mut self,
+        item: &ElementRc,
+        children_offset: u32,
+        parent_index: u32,
+        component_stack: &Self::SubComponentStack,
+    );
+    /// Called when a component is entered, this allow to change the component_stack.
+    /// The returned SubComponentStack will be used for all the items within that component
+    fn enter_component(
+        &mut self,
+        item: &ElementRc,
+        children_offset: u32,
+        component_stack: &Self::SubComponentStack,
+    ) -> Self::SubComponentStack;
+}
+
 /// Visit each item in order in which they should appear in the children tree array.
-/// The parameter of the visitor are
-///  1. The application specific state (can be used to keep track of how to reach fields)
-///  2. The current component being built (can be sub-component)
-///  3. the item
-///  4. the first_children_offset,
-///  5. the parent index
-///
-/// The visit_sub_component callback is called when encountering a sub-component.
-/// The parameters are:
-///  1. The application specific state
-///  2. The current component being built (the parent of the sub-component!)
-///  3. The element used to instantiate the sub-component
-///  4. The first_children_offset
 #[allow(dead_code)]
-pub fn build_item_tree<ComponentState>(
+pub fn build_item_tree<T: ItemTreeBuilder>(
     root_component: &Rc<Component>,
-    initial_state: ComponentState,
-    mut visit_item: impl FnMut(&ComponentState, &Rc<Component>, &ElementRc, u32, u32),
-    mut visit_sub_component: impl FnMut(
-        &ComponentState,
-        &Rc<Component>,
-        &ElementRc,
-        u32,
-    ) -> ComponentState,
+    initial_stack: &T::SubComponentStack,
+    helper: &mut T,
```

Comment:
I suggest to call this parameter `builder` (since it must implement `ItemTreeBuilder`).